File: go-algorithm/leetcode/problem380.go
package leetcode

import "math/rand"

// 380. Insert Delete GetRandom O(1)
// https://leetcode.com/problems/insert-delete-getrandom-o1/

type RandomizedSet struct {
	table map[int]int
	store []int
}

func _() RandomizedSet {
	return RandomizedSet{
		table: make(map[int]int),
		store: make([]int, 0),
	}
}

func (t *RandomizedSet) Insert(val int) bool {
	if _, ok := t.table[val]; ok {
		return false
	} else {
		t.store = append(t.store, val)
		t.table[val] = len(t.store) - 1
		return true
	}
}

func (t *RandomizedSet) Remove(val int) bool {
	if index, ok := t.table[val]; ok {
		t.store[index], t.store[len(t.store)-1] = t.store[len(t.store)-1], t.store[index]
		t.store = t.store[:len(t.store)-1]
		delete(t.table, val)
		if index < len(t.store) {
			t.table[t.store[index]] = index
		}
		return true
	} else {
		return false
	}
}

func (t *RandomizedSet) GetRandom() int {
	return t.store[rand.Intn(len(t.store))]
}

File: java-algorithm/src/main/java/com/leetcode/Problem380.java
package com.leetcode;

import java.util.*;

/**
 * 380. Insert Delete GetRandom O(1)
 * https://leetcode.com/problems/insert-delete-getrandom-o1/
 */
public class Problem380 {
    private static class RandomizedSet {
        private List<Integer> list;
        private Map<Integer, Integer> map;  // key: num, value: index

        public RandomizedSet() {
            this.list = new ArrayList<>();
            this.map = new HashMap<>();
        }

        public boolean insert(int val) {
            if (map.containsKey(val)) {
                return false;
            }
            list.add(val);
            map.put(val, list.size() - 1);
            return true;
        }

        public boolean remove(int val) {
            if (!map.containsKey(val)) {
                return false;
            }
            int index = map.get(val);
            list.set(index, list.get(list.size() - 1));
            list.remove(list.size() - 1);
            map.remove(val);
            if (index < list.size()) {
                map.put(list.get(index), index);
            }
            return true;
        }

        public int getRandom() {
            Random r = new Random();
            int pos = r.nextInt(list.size());
            return list.get(pos);
        }
    }
}
